2017/proposals/bill-engels-pymc3-proposal.md

```diff
@@ -0,0 +1,292 @@
+to refine or narrow scope if they deem necessary.  For instance, the scope
+could be narrowed to test one of the computation engines instead of both.
+
+- A series of blog posts comparing evaluations of PyMC3 on Theano, TensorFlow,
+and PyTorch 
+
+- Developer and PyMC3 architecture guide
+
+- Develop interface between PyMC3 and Theano
+
+## Schedule of Deliverables
+
+**Schedule is tentative**
+
+### May 1th - May 28th, **Community Bonding Period**
+
+As I get to know the mentors and other members of the PyMC3 community, I will
+study how to disentangle PyMC3 from Theano and research TensorFlow and PyTorch
+internals and functionality.
+
+- Experiment with and research architecture of Theano, TensorFlow, PyTorch, and PyMC3
+
+- Decide with mentors on the first alternate computation engine to try
+
+Prior to the community bonding period, I plan to continue submitting
+contributions to extend PyMC3's Gaussian Process functionality.
+
+### May 29th - June 3rd
+
+- Work through PyMC3 identifying and categorizing Theano calls as being either
+mathematical operations (`tensor.dot`, `tensor.transpose`, etc.) or
+Theano-specific (`theano.function`, `scan`, `tensor.as_tensor_variable`, etc.)
+
+- Check for glaring differences across backends in common operations such
+as array slicing behavior and the signatures of mathematical operations
+
+### June 5th - June 9th
+
+- Identify example model or existing unittests that would make good test cases
+
+- Begin porting the subset of PyMC3 needed to run the decided upon tests
+
+- Write a blog post comparing and contrasting TensorFlow, Theano, and PyTorch
+with a focus on their usage as general purpose computional engines, not as
+deep learning frameworks
+
+### June 12th - June 16th
+
+- Continue porting
+
+- Begin setting up the testing suite
+
+- Start draft of PyMC3 architecture and developer guide
+
+### June 19th - June 23th, **End of Phase 1**
+
+- Finish the first fork
+
+- Finish testing suite
+
+- Make sure the tests run
+
+### June 26 - June 30th, **Begin of Phase 2**
+
+- Document the comparison of the performance of Theano-PyMC3 and the
+first fork on the tests as a blog post
+
+### July 3rd - July 7th
+
+- Begin the second fork
+
+- Iterate on PyMC3 developer guide 
+
+### July 10th - July 14th
+
+- Continue work on the second fork
+
+### July 17th - July 21th, **End of Phase 2**
+
+- Finish the second fork
+
+- Make sure tests run
+
+### July 24th - July 28th, **Begin of Phase 3**
+
+- Compare performances
+
+- Understand why some models may work better than others in different forks
+
+- Document the comparison of the performance of Theano-PyMC3 and the
+second fork on the tests as a blog post
+
+### July 31st - August 4th
+
+- Test forks with GPU enabled
+
+- Write blog post comparing results on GPU
+
+### August 7th - August 11th
+
+- Finish up testing
+
+- Document overall results in a blog post
+
+- Design abstraction layer, or interface, between PyMC3 and Theano
```

Comment:
Yes, true.  What to do next will depend on results.  When writing the proposal I speculated that this would be a good next step, assuming that the most likely outcome is PyMC3 sticking with Theano.